contributors/design-proposals/architecture/namespaces.md

```diff
@@ -191,6 +191,52 @@ If the *namespace controller* sees a *Namespace* whose
 list is empty, it will signal the server to permanently remove the *Namespace*
 from storage by sending a final DELETE action to the API server.
 
+There are situations where the *namespace controller* is unable to guarantee 
+cleanup of all resources. During a cleanup run, it attempts a best-effort 
+resource deletion, remembers the errors that occurred and report back via
+**namespace status condition**. Some errors can be transient and will
+auto-resolve in the following cleanup runs, others may require manual
+intervention.
+
+There are following status conditions reporting on the process of namespace
+termination:
+- `NamespaceDeletionDiscoveryFailure` reports on errors during the first phase
+   of namespace termination - 
+   [resource discovery](../api-machinery/api-group.md).
+- `NamespaceDeletionGroupVersionParsingFailure` reports on errors that happen
+   when parsing the [GVK](../api-machinery/api-group.md) 
+   of all discovered resources.
+- `NamespaceDeletionContentFailure` reports on errors preventing the controller
+   to delete resources belonging to successfully discovered and parsed GVK.
```

Comment:
"preventing the controller to delete resources belonging" might be better as "preventing the controller from deleting resources belonging